src/main/java/com/commercetools/sync/shoppinglists/ShoppingListSync.java

```diff
@@ -0,0 +1,300 @@
+                    shoppingListDrafts.stream().map(ShoppingListDraft::getKey).collect(toSet());
+
+                return shoppingListService
+                    .fetchMatchingShoppingListsByKeys(shoppingListDraftKeys)
+                    .handle(ImmutablePair::new)
+                    .thenCompose(fetchResponse -> {
+                        final Set<ShoppingList> fetchedShoppingLists = fetchResponse.getLeft();
+                        final Throwable exception = fetchResponse.getRight();
+
+                        if (exception != null) {
+                            final String errorMessage = format(CTP_SHOPPING_LIST_FETCH_FAILED, shoppingListDraftKeys);
+                            handleError(new SyncException(errorMessage, exception), shoppingListDraftKeys.size());
+                            return CompletableFuture.completedFuture(null);
+                        } else {
+                            return syncBatch(fetchedShoppingLists, shoppingListDrafts);
+                        }
+                    });
+            })
+            .thenApply(ignoredResult -> {
+                statistics.incrementProcessed(batch.size());
+                return statistics;
+            });
+    }
+
+    @Nonnull
+    private CompletionStage<Void> syncBatch(@Nonnull final Set<ShoppingList> oldShoppingLists,
+                                            @Nonnull final Set<ShoppingListDraft>  newShoppingListDrafts) {
+
+        Map<String, ShoppingList> keyShoppingListsMap = oldShoppingLists
+            .stream()
+            .collect(toMap(ShoppingList::getKey, identity()));
+
+        return CompletableFuture.allOf(newShoppingListDrafts
+            .stream()
+            .map(shoppingListDraft ->
+                shoppingListReferenceResolver
+                    .resolveReferences(shoppingListDraft)
+                    .thenCompose(resolvedShoppingListDraft -> syncDraft(keyShoppingListsMap, resolvedShoppingListDraft))
+                    .exceptionally(completionException -> {
+                        final String errorMessage = format(FAILED_TO_PROCESS, shoppingListDraft.getKey(),
+                            completionException.getMessage());
+                        handleError(new SyncException(errorMessage, completionException), 1);
+                        return null;
+                    })
+            )
+            .map(CompletionStage::toCompletableFuture)
+            .toArray(CompletableFuture[]::new));
+    }
+
+    @Nonnull
+    private CompletionStage<Void> syncDraft(@Nonnull final Map<String, ShoppingList> keyShoppingListMap,
+                                            @Nonnull final ShoppingListDraft newShoppingListDraft) {
+
+        final ShoppingList oldShoppingList = keyShoppingListMap.get(newShoppingListDraft.getKey());
+        return Optional.ofNullable(oldShoppingList)
+                       .map(customer -> buildActionsAndUpdate(oldShoppingList, newShoppingListDraft))
```

Comment:
`customer`? 🤔

Reply:
why we ignore that parameter?

Reply:
Typo. Changed to `shoppingList`
Also, the parameter ignoring seems to be 'tradition' in many other resources. I just fixed it in `ShoppingListSync`.
https://github.com/commercetools/commercetools-sync-java/pull/615/commits/05c709103f58ecc8c9b1d3b5ebc2bdf0bcf52e98

Reply:
I'd rename it to `oldShoppingList` for more clarity.

Reply:
@JudeNiroshan but the variable naming duplicates with existing `oldShoppingList`.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/615/commits/dd0d20797b9c91c222b1cd1b854425555e300885